### PR TITLE
Role rank

### PR DESCRIPTION
**Pull request by Stankman**
Modified roles' routes sort:

### List Route
`/list/{page}/{limit}`
`/permissions/list/{page}/{limit}`
`/view/my-permissions`

### Creation Routes
`/create/role`
`/create/permission`

### Assign Routes
`/assign/rank`
`/assign/child`
`/assign/permission`

### Delete Routes
`/delete/role`
`/delete/permission`

---

**With this modification: Roles now work per Rank. Instead of assigning 2 times a user level, now it uses a single assignation whether `via role` or `via rank`.**

**Role Hierarchy** is now limited to 3 levels. (Possible to change this in the future)

Level 1: `Root Level`
Level 2: `Group Level`
Level 3: `Rank Level`

### CODE REVIEW

File: src/Role/Repository/RoleHierarchyRepository.php [hunk 1]
```diff
@@ -90,6 +108,71 @@ public function hasChildRoleId(int $parentRoleId, int $findingChildId): bool
...
+                if($this->hasParentRoleId($parentId, $findingParentId)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    public function areBrothers(int $parentRoleId, int $childRoleId) : bool {
```
Comment (Domexx) [inline src/Role/Repository/RoleHierarchyRepository.php:148]: changes are looking good please add a docblock over the function that describes what it does

File: src/Role/Repository/RoleHierarchyRepository.php [hunk 2]
```diff
@@ -90,6 +108,71 @@ public function hasChildRoleId(int $parentRoleId, int $findingChildId): bool
...
+                if($this->areBrothers($parentId, $parentRoleId)) {
+                    return true;
+                }
+            }
+        }
+        
+        return false;
+    }
+
+    public function roleIsGrandChild(int $parentRoleId, $parentsCount = 0) : bool {
```
Comment (Domexx) [inline src/Role/Repository/RoleHierarchyRepository.php:174]: like here, describe the function and add a docblock

File: src/Role/Repository/RoleHierarchyRepository.php [hunk 3]
```diff
@@ -37,9 +37,27 @@ private function getChildIds(array $parentIds): ?array
...
+     * @param array $childIds
+     *
+     * @return array|null
+     */
+    private function getParentIds(array $childIds): ?array
+    {
+        $searchCriteria = $this->getDataObjectManager()
+            ->whereIn('child_role_id', $childIds);
+
+        if($searchCriteria->count() == 0) return [];
```
Comment (Domexx) [inline src/Role/Repository/RoleHierarchyRepository.php]: we dont use singleline ifs, please write it out to the fully

if() {
return stuff happens here
}

File: src/Role/Repository/RoleHierarchyRepository.php [hunk 4]
```diff
@@ -37,9 +37,27 @@ private function getChildIds(array $parentIds): ?array
         $searchCriteria = $this->getDataObjectManager()
             ->whereIn('parent_role_id', $parentIds);
 
+        if($searchCriteria->count() == 0) return [];
```
Comment (Domexx) [inline src/Role/Repository/RoleHierarchyRepository.php]: we dont use singleline ifs, please write it out to the fully

if() {
return stuff happens here
}

File: src/Role/Service/FetchUserPermissionService.php [hunk 5]
```diff
@@ -34,24 +34,21 @@ public function __construct(
...
-        $userRoleIds = $this->roleUserRepository->getUserRoleIds($userId);
-
-        if (!$userRoleIds) {
-            return response()
-                ->setData([]);
+        /** @var array $rankRoleIds */
+        $rankRoleIds = $this->roleRankRepository->getRankRoleIds($rank);
+        
+        if (!$rankRoleIds) {
+            return response()->setData([]);
```
Comment (Domexx) [inline src/Role/Service/FetchUserPermissionService.php:43]: We could probably throw an error here? Is it needed to return with an empty array? Maybe respond with a custom error code for the frontend?

Are there ranks without roles?

**Review by Domexx — CHANGES_REQUESTED** [review]
Please look at the comments I made on the code, let me know if you have any questions

File: src/Role/Service/FetchUserPermissionService.php (hunk 5 above)
Comment (Stankman) [reply src/Role/Service/FetchUserPermissionService.php:43]: Well, rank 1 would be a rank without role

Comment (Stankman) [reply src/Role/Service/FetchUserPermissionService.php:43]: And there would be also Roles without Ranks just as "Root" Level or "Group" level Roles

**Review by Domexx — APPROVED** [review]
Looks good now :)

**Pull request merged**